google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/util/AppEngineUtil.java

```diff
@@ -16,13 +16,206 @@
 
 package com.google.cloud.tools.intellij.appengine.util;
 
+import com.google.cloud.tools.intellij.appengine.cloud.AppEngineArtifactDeploymentSource;
+import com.google.cloud.tools.intellij.appengine.cloud.AppEngineEnvironment;
+import com.google.cloud.tools.intellij.appengine.cloud.MavenBuildDeploymentSource;
+import com.google.cloud.tools.intellij.appengine.cloud.UserSpecifiedPathDeploymentSource;
+import com.google.common.collect.Lists;
+
+import com.intellij.facet.Facet;
+import com.intellij.facet.FacetManager;
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.module.ModuleManager;
+import com.intellij.openapi.module.ModulePointer;
+import com.intellij.openapi.module.ModulePointerManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.packaging.artifacts.Artifact;
+import com.intellij.packaging.artifacts.ArtifactManager;
+import com.intellij.packaging.artifacts.ArtifactPointerManager;
+import com.intellij.packaging.impl.artifacts.ArtifactUtil;
+import com.intellij.remoteServer.configuration.deployment.ModuleDeploymentSource;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.idea.maven.project.MavenProject;
+import org.jetbrains.idea.maven.project.MavenProjectsManager;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
 /**
  * App Engine utility methods.
  */
 public class AppEngineUtil {
 
+  private static final String APP_ENGINE_STANDARD_FACET_NAME = "Google App Engine";
+
   private AppEngineUtil() {
     // Not designed for instantiation
   }
 
+  /**
+   * Creates a list of artifact deployment sources available for deployment to App Engine.
+   *
+   * <p>Artifacts either target the standard or the flexible environment. All standard artifacts are
+   * added. Flexible artifacts are only added if there are no other standard artifacts associated
+   * with the same module - i.e. if a module is set up for App Engine standard (has an
+   * appengine-web.xml etc.) then no option is provided to deploy any of its artifacts to the
+   * flexible environment.
+   *
+   * @return a list of {@link AppEngineArtifactDeploymentSource}'s
+   */
+  public static List<AppEngineArtifactDeploymentSource> createArtifactDeploymentSources(
+      @NotNull Project project) {
+    List<AppEngineArtifactDeploymentSource> sources = Lists.newArrayList();
+
+    for (Module module : ModuleManager.getInstance(project).getModules()) {
```

Comment:
@etanshaul   I thinkm this is the last question I have before we merge this and iterate.